test/projection.jl

```diff
@@ -1,240 +1,161 @@
+        @test pvec3(1.0:3.0) === 1.0:3.0
+        @test pvec3(1:3) == 1.0:3.0  # would prefer ===, map(Float64, dx) would do that, not important
+        @test pvec3([1,2,3+4im]) == 1:3
+        @test eltype(pvec3([1,2,3f0])) === Float64
+
+        # reshape
+        @test pvec3(reshape([1,2,3],3,1)) isa Vector
+        @test_throws DimensionMismatch pvec3(reshape([1,2,3],1,3))
+        @test_throws DimensionMismatch pvec3([1,2,3,4])
+
+        pmat = ProjectTo(rand(2,2) .+ im)
+        @test pmat([1 2; 3 4.0 + 5im]') isa Adjoint     # pass-through
+        @test pmat([1 2; 3 4]') isa Matrix{ComplexF64}  # broadcast type change
+
+        # arrays of arrays
+        pvecvec = ProjectTo([[1,2], [3,4,5]])
+        @test pvecvec([1:2, 3:5])[1] == 1:2
+        @test pvecvec([[1,2+3im], [4+5im,6,7]])[2] == [4,6,7]
+        @test pvecvec(hcat([1:2, hcat(3:5)]))[2] isa Vector  # reshape inner & outer
+
+        # arrays of unknown things
+        @test ProjectTo([:x, :y])(1:2) === 1:2  # no element handling,
+        @test ProjectTo([:x, :y])(reshape(1:2,2,1,1)) == 1:2  # but still reshapes container
+        @test ProjectTo(Any[1, 2])(1:2) === 1:2  # goes by eltype, hence ignores contents.
     end
 
-    @testset "to Array" begin
-        # to an array of numbers
-        x = zeros(2, 2)
-        @test [1.0 2.0; 3.0 4.0] == ProjectTo(x)([1.0 2.0; 3.0 4.0])
-        @test x == ProjectTo(x)(ZeroTangent())
-
-        x = zeros(2)
-        @test x == ProjectTo(x)(@thunk(ZeroTangent()))
-
-        x = zeros(Float32, 2, 2)
-        @test x == ProjectTo(x)([0.0 0; 0 0])
-
-        x = [1.0 0; 0 4]
-        @test x == ProjectTo(x)(Diagonal([1.0, 4]))
-
-        # to a array of structs
-        x = [Fred(0.0), Fred(0.0)]
-        @test x == ProjectTo(x)([Fred(0.0), Fred(0.0)])
-        @test x == ProjectTo(x)([ZeroTangent(), ZeroTangent()])
-        @test x == ProjectTo(x)([ZeroTangent(), @thunk(Fred(0.0))])
-        @test x == ProjectTo(x)(ZeroTangent())
-        @test x == ProjectTo(x)(@thunk(ZeroTangent()))
-
-        x = [Fred(1.0) Fred(0.0); Fred(0.0) Fred(4.0)]
-        @test x == ProjectTo(x)(Diagonal([Fred(1.0), Fred(4.0)]))
-    end
-
-    @testset "To Array of Arrays" begin
-        # inner arrays have same type but different sizes
-        x = [[1.0, 2.0, 3.0], [4.0, 5.0]]
-        @test x == ProjectTo(x)(x)
-        @test x == ProjectTo(x)([[1.0 + 2im, 2.0, 3.0], [4.0 + 2im, 5.0]])
-
-        # This makes sure we don't fall for https://github.com/JuliaLang/julia/issues/38064
-        @test [[0.0, 0.0, 0.0], [0.0, 0.0]] == ProjectTo(x)(ZeroTangent())
+    @testset "Base: zero-arrays & Ref" begin
+        pzed = ProjectTo(fill(1.0))
+        @test pzed(fill(3.14)) == fill(3.14)  # easy
+        @test pzed(fill(3)) == fill(3.0)      # broadcast type change must not produce number
+        @test pzed(hcat(3.14)) == fill(3.14)  # reshape of length-1 array
+        @test pzed(3.14 + im) == fill(3.14)   # re-wrap of a scalar number
```

Comment:
Right now it re-wraps Number -> Array{T,0}. Maybe deserves discussion.

Reply:
I am sold